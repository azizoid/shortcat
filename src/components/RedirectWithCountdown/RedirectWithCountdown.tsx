import { useRouter } from 'next/router';
import React, { useEffect, useState } from 'react';

type RedirectWithCountdownProps = {
  redirectUrl: string
}

export const RedirectWithCountdown = ({ redirectUrl }: RedirectWithCountdownProps) => {
  const router = useRouter();
  const [countdown, setCountdown] = useState(5);

  useEffect(() => {
    const countdownInterval = setInterval(() => {
      setCountdown((prevCountdown) => prevCountdown - 1);
    }, 1000);

    setTimeout(() => {
      clearInterval(countdownInterval);
    }, 5000);

    return () => clearInterval(countdownInterval);
  }, [redirectUrl]);

  useEffect(() => {
    if (countdown === 0) {
      router.replace(redirectUrl);
    }
  }, [countdown, redirectUrl, router]);

  if (countdown === 0) {
    return null;
  }

  return (
    <div className="whitespace-nowrap font-medium text-center text-gray-900 dark:text-white">
      <p>You will be redirected to {redirectUrl} in {countdown} seconds.</p>
      <button onClick={() => router.push(redirectUrl)}>Click here if you are not automatically redirected.</button>
    </div>
  );
}