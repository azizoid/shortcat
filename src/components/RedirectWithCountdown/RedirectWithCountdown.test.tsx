import { RouterContext } from 'next/dist/shared/lib/router-context';
import { render, act } from '@testing-library/react';
import { RedirectWithCountdown } from './RedirectWithCountdown';

const mockRouter = {
  replace: jest.fn(),
  push: jest.fn(),
};

const renderRedirectWithCountdown = (redirectUrl: string) => {
  return render(
    <RouterContext.Provider value={{ ...mockRouter }}>
      <RedirectWithCountdown redirectUrl={redirectUrl} />
    </RouterContext.Provider>
  );
};

describe('RedirectWithCountdown', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should render a countdown and redirect after 5 seconds', () => {
    const redirectUrl = 'https://nam.az';
    const { container } = renderRedirectWithCountdown(redirectUrl);
    expect(container).toMatchSnapshot()
  });

  it('should redirect immediately if countdown is set to 0', () => {
    const redirectUrl = 'https://nam.az';
    const { queryByText } = renderRedirectWithCountdown(redirectUrl);
    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(mockRouter.replace).toHaveBeenCalledWith(redirectUrl);

    expect(queryByText(`You will be redirected to ${redirectUrl} in 0 seconds.`)).not.toBeInTheDocument();
  });

  it('should redirect when the button is clicked', () => {
    const redirectUrl = 'https://nam.az';
    const { getByRole } = renderRedirectWithCountdown(redirectUrl);

    const button = getByRole('button');
    expect(button).toBeInTheDocument();

    act(() => {
      button.click();
    });

    expect(mockRouter.push).toHaveBeenCalledWith(redirectUrl);
  });
});