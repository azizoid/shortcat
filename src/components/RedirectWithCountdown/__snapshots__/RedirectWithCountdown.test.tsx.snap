// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedirectWithCountdown should render a countdown and redirect after 5 seconds 1`] = `
<div>
  <div
    class="whitespace-nowrap font-medium text-center text-gray-900 dark:text-white"
  >
    <p>
      You will be redirected to 
      https://nam.az
       in 
      5
       seconds.
    </p>
    <button>
      Click here if you are not automatically redirected.
    </button>
  </div>
</div>
`;
